src/control/lib/hardware/sysfs/provider.go

```diff
@@ -241,3 +245,64 @@ func (s *Provider) getCXIFabricInterfaces() ([]*hardware.FabricInterface, error)
 
 	return cxiFIs, nil
 }
+
+// CheckFabricReady inspects the fabric interface to determine whether it is fully initialized
+// and ready for use.
+func (s *Provider) CheckFabricReady(iface string) error {
+	if s == nil {
+		return errors.New("sysfs provider is nil")
+	}
+
+	if iface == "" {
+		return errors.New("fabric interface name is required")
+	}
+
+	devClass, err := s.GetNetDevClass(iface)
+	if err != nil {
+		return errors.Wrapf(err, "can't determine device class for %q", iface)
+	}
+
+	if devClass != hardware.Infiniband {
+		// Infiniband is the only type we need to actively check
+		return nil
+	}
+
+	// The best way to determine that an Infiniband interface is ready is to check the state
+	// of its ports. Ports in the "ACTIVE" state are either fully ready or will be very soon.
+	ibPath := s.sysPath("class", "net", iface, "device", "infiniband")
+	ibDevs, err := ioutil.ReadDir(ibPath)
+	if err != nil {
+		return errors.Wrapf(err, "can't access Infiniband details for %q", iface)
+	}
+
+	for _, dev := range ibDevs {
+		portPath := filepath.Join(ibPath, dev.Name(), "ports")
+		ports, err := ioutil.ReadDir(portPath)
+		if err != nil {
+			return errors.Wrapf(err, "unable to get ports for %s/%s", iface, dev.Name())
+		}
+
+		for _, port := range ports {
+			statePath := filepath.Join(portPath, port.Name(), "state")
+			stateBytes, err := ioutil.ReadFile(statePath)
+			if err != nil {
+				return errors.Wrapf(err, "unable to get state for %s/%s port %s",
+					iface, dev.Name(), port.Name())
+			}
+
+			stateStrs := strings.Split(string(stateBytes), ": ")
+			stateNum, err := strconv.Atoi(stateStrs[0])
+			if err != nil {
+				s.log.Debugf("unable to parse %s/%s port %s state %q: %s",
+					iface, dev.Name(), port.Name(), string(stateBytes), err.Error())
+				continue
+			}
+			if stateNum == ibPortActiveState {
+				// At least one IB port is active on the interface
```

Comment:
Hmm. Don't want to make the solution more complicated than necessary, but I wonder if we really should be checking that a specific port is active rather than just being OK with any being active? I will readily admit that I don't know all of the details here, but my naive take on this is that it might be inviting more undefined behavior.

Reply:
Yeah, I'm not sure how we would select the port based on the DAOS configuration though. We don't know anything about that at the DAOS level, and I'm not totally sure how we'd figure it out before the card is fully up and running, with the domain detected.

Reply:
OK, fair enough.